File: datasets/Part 2 - Regression/Section 6 - Polynomial Regression/polinomial_regression_sbn.R
# Regresion polinomica
# Importar dataset
dataset <- read.csv('Position_Salaries.csv')
dataset = dataset[, 2:3]

# Dividir el dataset en entrenamiento y prueba
#install.packages('caTools')
# library(caTools)
# set.seed(123) SOLO ES PARA QUE SIEMPRE SELECCIONE LOS MISMOS ELEMENTOS
# split <- sample.split(dataset$Purchased, SplitRatio = 0.8)
# trainning_set <- subset(dataset, split == TRUE)
# testing_set <- subset(dataset, split == FALSE)

# Escalado de variables
# trainning_set[, 2:3] <- scale(trainning_set[, 2:3])
# testing_set[, 2:3] <- scale(testing_set[, 2:3])+

# Ajustar la regresion lineal con el dataset
lineal_reg <- lm(formula = Salary ~ .,
                 data = dataset)
# summary(lineal_reg)

# Ajustar la regresion polinomica con el dataset
dataset$Level2 <- dataset$Level ^ 2
dataset$Level3 <- dataset$Level ^ 3
dataset$Level4 <- dataset$Level ^ 4
poly_reg <- lm(formula = Salary ~ .,
                 data = dataset)
# summary(poly_reg)

# Visualizacion de los resultados del modelo lineal
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level,
                 y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level,
                y = predict(lineal_reg,
                            newdata = dataset)),
            colour = 'blue') +
  ggtitle('Modelo de Regresion Lineal') +
  xlab('Posición del empleado') +
  ylab('Sueldo en USD $')

# Visualizacion de los resultados del modelo polinomico
x_grid <- seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level,
                 y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid,
                y = predict(poly_reg,
                            newdata = data.frame(Level = x_grid,
                                                 Level2 = x_grid ^ 2,
                                                 Level3 = x_grid ^ 3,
                                                 Level4 = x_grid ^ 4))),
            colour = 'blue') +
  ggtitle('Modelo de Regresion') +
  xlab('Posición del empleado') +
  ylab('Sueldo en USD $')

# Prediccion de nuestros modelos
y_pred <- predict(lineal_reg,
                   newdata = data.frame(Level = 6.5))

y_pred <- predict(poly_reg,
                  newdata = data.frame(Level = 6.5,
                                       Level2 = 6.5 ^ 2,
                                       Level3 = 6.5 ^ 3,
                                       Level4 = 6.5 ^ 4))

File: datasets/Part 2 - Regression/Section 6 - Polynomial Regression/polinomial_regression_sbn.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jan 10 15:29:53 2020

@author: jguerrero
"""

# Regresion polinomica
# Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('Position_Salaries.csv')

X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Dividir el dataset en entrenamiento y prueba
'''from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)#, random_state = 0) SOLO ES PARA QUE SIEMPRE SELECCIONE LOS MISMOS ELEMENTOS'''

# Escalado de variables
# Estandarizacion
'''from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
# Escalar los datos de train y test con la misma formula
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)'''

# Ajustar la regresion lineal con el dataset
from sklearn.linear_model import LinearRegression
lineal_reg = LinearRegression()
lineal_reg.fit(X, y)

# Ajustar la regresion polinomica con el dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
#X_poly = poly_reg.fit_transform(X)
lineal_reg_2 = LinearRegression()
#lineal_reg_2.fit(X_poly, y)
lineal_reg_2.fit(poly_reg.fit_transform(X), y)

# Visualizacion de los resultados del modelo lineal
plt.scatter(X, y, color = 'red')
plt.plot(X, lineal_reg.predict(X), color = 'blue')
plt.title('Modelo de Regresion Lineal')
plt.xlabel('Posición del empleado')
plt.ylabel('Sueldo en USD $')
plt.show()

# Visualizacion de los resultados del modelo polinomico
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = 'red')
#plt.plot(X, lineal_reg_2.predict(X_poly), color = 'blue')
plt.plot(X_grid, lineal_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Modelo de Regresion Polinomica')
plt.xlabel('Posición del empleado')
plt.ylabel('Sueldo en USD $')
plt.show()

# Prediccion de nuestros modelos
lineal_reg.predict([[6.5]])
lineal_reg_2.predict(poly_reg.fit_transform([[6.5]]))